### PR TITLE
resource id should be readonly. space name on datasource should be optional

### PR DESCRIPTION
**Pull request by benPearce1**
setup:

```
resource "octopusdeploy_lifecycle" "lifecycle" {
  id = "Lifecycles-1"
  space_id = data.octopusdeploy_space.space.id
  name = "lifecycle 2"
}
```

on `0.22.1`
```
# octopusdeploy_lifecycle.lifecycle will be created
  + resource "octopusdeploy_lifecycle" "lifecycle" {
      + description = "description2"
      + id          = (known after apply)
      + name        = "lifecycle 2"
      + space_id    = "Spaces-1"
...
```
Lifecycle resource was created with id of `Lifecycles-1762`. From state:
```
"id": "Lifecycles-1762",
"name": "lifecycle 2",
```

Plan with `0.30.0-beta8`
```
# octopusdeploy_lifecycle.lifecycle will be updated in-place
  ~ resource "octopusdeploy_lifecycle" "lifecycle" {
      ~ id          = "Lifecycles-1762" -> "Lifecycles-1"
        name        = "lifecycle 2"
        # (2 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }
```

Plan now fails validation as Id is readonly:
```
│ Error: Invalid Configuration for Read-Only Attribute
│ 
│   with octopusdeploy_lifecycle.lifecycle,
│   on main.tf line 32, in resource "octopusdeploy_lifecycle" "lifecycle":
│   32:   id = "Lifecycles-1"
│ 
│ Cannot set value for this attribute as the provider has marked it as read-only. Remove the configuration line setting the value.
│ 
│ Refer to the provider documentation or contact the provider developers for additional information about configurable and read-only attributes that are
│ supported.
╵
```

### CODE REVIEW

File: octopusdeploy_framework/schemas/schema.go [hunk 1]
```diff
@@ -130,7 +130,6 @@ func GetIdResourceSchema() resourceSchema.Attribute {
 	return resourceSchema.StringAttribute{
 		Description: "The unique ID for this resource.",
 		Computed:    true,
-		Optional:    true,
```
Comment (benPearce1) [inline octopusdeploy_framework/schemas/schema.go:133]: This means that the user cannot set the Id and will get an error

File: octopusdeploy_framework/schemas/space.go [hunk 2]
```diff
@@ -69,14 +69,15 @@ func (s SpaceSchema) GetDatasourceSchema() datasourceSchema.Schema {
...
-			"id":          GetIdDatasourceSchema(true),
+			"id":          GetIdDatasourceSchema(false),
 			"description": GetReadonlyDescriptionDatasourceSchema(spaceDescription),
 			"name": datasourceSchema.StringAttribute{
 				Description: fmt.Sprintf("The name of this resource, no more than %d characters long", 20),
 				Validators: []validator.String{
 					stringvalidator.LengthBetween(1, 20),
 				},
 				Computed: true,
+				Optional: true,
```
Comment (benPearce1) [inline octopusdeploy_framework/schemas/space.go:80]: Looks like this was broken from #759

File: octopusdeploy_framework/schemas/variable.go [hunk 3]
```diff
@@ -148,10 +148,7 @@ func (v VariableSchema) GetResourceSchema() resourceSchema.Schema {
 	return resourceSchema.Schema{
 		Description: util.GetResourceSchemaDescription(VariableResourceDescription),
 		Attributes: map[string]resourceSchema.Attribute{
-			SchemaAttributeNames.ID: resourceSchema.StringAttribute{
-				Description: "The ID of this resource.",
-				Computed:    true,
-			},
+			SchemaAttributeNames.ID:          GetIdResourceSchema(),
```
Comment (benPearce1) [inline octopusdeploy_framework/schemas/variable.go:151]: Use the helper function for consistency

**Review by domenicsim1 — APPROVED** [review]
Good catch 🪝

**Pull request merged**